### PR TITLE
bgpv2: Adding pod cidr advertisement reconciler for BGPv2 APIs

### PR DESCRIPTION
**Pull request by harsimran-pabla**
Introducing pod cidr advertisement reconciler in BGP v2 reconcilers. 

- BGPv1 code change - adding option to set family field in path attributes. If no family is set, behavior is unchanged.
- Introducing new peer_advertisement cell inside reconcilers. This can be used by various advertisement reconcilers to get configured advertisement - per peer per family.
- Advertisement reconciler - generic layer to advertise/withdraw prefixes from BGP instance.
- PodCIDR reconciler - this uses all of the above to manage pod cidr advertisements.

```release-note
  bgpv2: Introducing pod cidr reconciler for bgpv2.
```

### CODE REVIEW

File: pkg/bgpv1/gobgp/conversions.go [hunk 1]
```diff
@@ -96,6 +101,71 @@ func ToAgentPaths(paths []*gobgp.Path) ([]*types.Path, error) {
...
 }
 
+func toGoBGPFamily(family types.Family) *gobgp.Family {
+	return &gobgp.Family{
+		Afi:  toGoBGPAfi(family.Afi),
+		Safi: toGoBGPSafi(family.Safi),
+	}
+}
+
+func toGoBGPAfi(a types.Afi) gobgp.Family_Afi {
```
Comment (rastislavs) [inline pkg/bgpv1/gobgp/conversions.go:111]: Inverse functions to these (`toAgentAfi`, `toAgentSafi`) are located in `gobgp/state.go` ATM, maybe this would be a good opportunity to consolidate and move them into `conversions.go` as well?

File: pkg/bgpv1/manager/reconcilerv2/peer_advertisements.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+func NewCiliumPeerAdvertisement(p PeerAdvertisementIn) *CiliumPeerAdvertisement {
+	return &CiliumPeerAdvertisement{
+		logger:     p.Logger,
+		peerConfig: p.PeerConfigStore,
+		adverts:    p.AdvertStore,
+	}
+}
+
+// GetAdvertisements can be called to get advertisements for given BGPAdvertisementType.
```
Comment (rastislavs) [inline pkg/bgpv1/manager/reconcilerv2/peer_advertisements.go]:
```suggestion
// GetAdvertisements can be called to get all configured advertisements of given BGPAdvertisementType for each peer.
```

File: pkg/bgpv1/manager/reconcilerv2/advertisements.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+				return nil, err
+			}
+		}
+
+		return nil, nil
+	}
+
+	// an aset member which book keeps which universe it exists in
+	type member struct {
+		a     bool
```
Comment (rastislavs) [inline pkg/bgpv1/manager/reconcilerv2/advertisements.go]: What abut renaming `a` to `configured` and `b` to `advertised`? That would make the code a bit more readable.

File: pkg/bgpv1/types/state.go [hunk 4]
```diff
@@ -254,3 +258,10 @@ func ParseSafi(s string) Safi {
 	}
 	return ret
 }
+
+func ToAgentFamily(fam v2alpha1.CiliumBGPFamily) Family {
```
Comment (rastislavs) [inline pkg/bgpv1/types/state.go:262]: Maybe a good opportunity to rename this file to `conversions.go`, as it is (already) used not only for state data retrieval?

File: pkg/bgpv1/manager/reconcilerv2/pod_cidr.go [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+			if ok {
+				for _, existingPath := range existingPaths {
+					existingPathsSet.Insert(existingPath.NLRI.String())
+				}
+			}
+
+			var paths []*types.Path
+			for _, prefix := range desiredPrefixes {
+				path := types.NewPathForPrefix(prefix)
+				path.UpdateFamilyField(agentFamily)
```
Comment (rastislavs) [inline pkg/bgpv1/manager/reconcilerv2/pod_cidr.go]: What is the benefit of using the `UpdateFamilyField` helper instead of just `path.Family = agentFamily`?

File: pkg/bgpv1/manager/reconcilerv2/pod_cidr.go [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+				paths = append(paths, path)
+			}
+
+			desiredFamilyAdverts[types.ToAgentFamily(family)] = append(desiredFamilyAdverts[types.ToAgentFamily(family)], paths...)
+		}
+	}
+	return desiredFamilyAdverts, nil
+}
+
+func (r *ExportPodCIDRReconciler) reconcilerFamilyAdvertisements(ctx context.Context, p ReconcileParams, desiredFamilyAdverts ExportPodCIDRMetadata) error {
```
Comment (rastislavs) [inline pkg/bgpv1/manager/reconcilerv2/pod_cidr.go]:
```suggestion
func (r *ExportPodCIDRReconciler) reconcileFamilyAdvertisements(ctx context.Context, p ReconcileParams, desiredFamilyAdverts ExportPodCIDRMetadata) error {
```

File: pkg/bgpv1/manager/reconcilerv2/pod_cidr.go [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+			Enabled:               true,
+			Instance:              p.BGPInstance,
+			InstanceName:          p.DesiredConfig.Name,
+			CurrentAdvertisements: runningFamilyAdverts[family],
+			ToAdvertise:           desiredFamilyAdverts[family],
+		})
+		if err != nil {
+			return err
+		}
+		runningFamilyAdverts[family] = runningAdverts
```
Comment (rastislavs) [inline pkg/bgpv1/manager/reconcilerv2/pod_cidr.go]: Could we do this in a single `ExportAdvertisementsReconciler` call per family?

File: pkg/bgpv1/manager/reconcilerv2/advertisements.go [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+		toAdvertise []*types.Path
+		// holds advertisements which must remain in place
+		toKeep []*types.Path
+		// holds advertisements which must be removed
+		toWithdraw []*types.Path
+		// the result of advertising toAdvertise.
+		newAdverts []*types.Path
+	)
+
+	l.Debug("begin reconciling advertisements for BGP instance")
```
Comment (rastislavs) [inline pkg/bgpv1/manager/reconcilerv2/advertisements.go]: Isn't this a bit too verbose logging, given that we may call this multiple times per reconciler for each reconcile? Maybe we should log only when adding new advertisements / withdrawing paths?

File: pkg/bgpv1/manager/reconcilerv2/peer_advertisements.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+		if err != nil {
+			return result, err
+		}
+		result[family.CiliumBGPFamily] = advert
+	}
+	return result, nil
+}
+
+func (p *CiliumPeerAdvertisement) getFamilyAdvertisements(family v2alpha1.CiliumBGPFamilyWithAdverts, selectAdvertTypes ...v2alpha1.BGPAdvertisementType) ([]v2alpha1.Advertisement, error) {
+	// get all advertisement CRDs
```
Comment (rastislavs) [inline pkg/bgpv1/manager/reconcilerv2/peer_advertisements.go]: nit: all CRD objects, there is only one advertisement CRD

File: pkg/bgpv1/manager/reconcilerv2/peer_advertisements.go [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+// GetAdvertisements can be called to get advertisements for given BGPAdvertisementType.
+// Advertisements are selected based on below criteria:
+// Each peer is selected from the BGP node instance configuration. For each peer, the peer configuration is fetched
+// from local store.
+// Peer configuration contains the list of families and the advertisement selector.
+// We iterate over all advertisements ( available from local store ), select only those that match the advertisement
+// selector of the family.
+// Information of peer -> family -> advertisements is returned to the consumer.
+func (p *CiliumPeerAdvertisement) GetAdvertisements(conf *v2alpha1.CiliumBGPNodeInstance, selectAdvertTypes ...v2alpha1.BGPAdvertisementType) (PeerAdvertisements, error) {
```
Comment (rastislavs) [inline pkg/bgpv1/manager/reconcilerv2/peer_advertisements.go]: nit: maybe I would name this `GetConfiguredAdvertisements` or similar, to make it obvious that this is returning configuration, not state data.

**Review by rastislavs — CHANGES_REQUESTED** [review]
Thanks, very nice work! Left some comments, but the overall approach looks good to me.

File: pkg/bgpv1/manager/reconcilerv2/pod_cidr.go (hunk 5 above)
Comment (harsimran-pabla) [reply pkg/bgpv1/manager/reconcilerv2/pod_cidr.go]: I was thinking `path := types.NewPathForPrefix(prefix).PathWithFamily(agentFamily)`, somewhere in middle gave up on that idea. 

I'll go with `path.Family = agentFamily`, seems clear enough.

File: pkg/bgpv1/manager/reconcilerv2/pod_cidr.go (hunk 7 above)
Comment (harsimran-pabla) [reply pkg/bgpv1/manager/reconcilerv2/pod_cidr.go]: I see there are two cases - delete, upsert. Both are called with different parameters of ExportAdvertisementReconciler; like enabled field. 

We could make a single call, but then deleted families might have stale family field with empty paths in metadata.
`runningFamilyAdverts[family] = nil`. I wanted to remove deleted family completely from runningFamilyAdverts map.

Comment (rastislavs) [reply pkg/bgpv1/manager/reconcilerv2/pod_cidr.go]: ACK - understood, thx

File: pkg/bgpv1/manager/reconcilerv2/advertisements.go [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+
+// ExportAdvertisementsReconciler reconciles the state of the BGP advertisements
+// with the provided toAdvertise list and returns a list of the advertisements
+// currently being announced.
+func ExportAdvertisementsReconciler(params *AdvertisementsReconcilerParams) ([]*types.Path, error) {
+	var (
+		// logger for the reconciler
+		l = params.Logger.WithFields(logrus.Fields{
+			"reconciler": params.Reconciler,
+			"instance":   params.InstanceName,
```
Comment (YutaroHayakawa) [inline pkg/bgpv1/manager/reconcilerv2/advertisements.go]: Let's use logfield.

File: pkg/bgpv1/manager/reconcilerv2/pod_cidr.go [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+	return ExportPodCIDRReconcilerOut{
+		Reconciler: &ExportPodCIDRReconciler{
+			logger:     params.Logger.WithField(types.ReconcilerLogField, "PodCIDR"),
+			peerAdvert: params.PeerAdvert,
+		},
+	}
+}
+
+func (r *ExportPodCIDRReconciler) Name() string {
+	return "ExportPodCIDR"
```
Comment (YutaroHayakawa) [inline pkg/bgpv1/manager/reconcilerv2/pod_cidr.go]: Shall we take this opportunity to change the name to PodCIDR instead of ExportPodCIDR? I was feeling a bit weird about `Export` prefix because rest of the reconcilers also exports something, but not prefixed with `Export` (e.g., LBService).

File: pkg/bgpv1/manager/reconcilerv2/pod_cidr.go [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+
+type ExportPodCIDRReconciler struct {
+	logger     logrus.FieldLogger
+	peerAdvert *CiliumPeerAdvertisement
+}
+
+// ExportPodCIDRMetadata is a map of advertisements per family, key is family type
+type ExportPodCIDRMetadata map[types.Family][]*types.Path
+
+func NewExportPodCIDRReconciler(params ExportPodCIDRReconcilerIn) ExportPodCIDRReconcilerOut {
```
Comment (YutaroHayakawa) [inline pkg/bgpv1/manager/reconcilerv2/pod_cidr.go]: Should we check the IPAM type here and not provide this reconciler for unsupported IPAM modes? We can also do it in the separated PR.

File: pkg/bgpv1/manager/reconcilerv2/advertisements.go [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+		newAdverts []*types.Path
+	)
+
+	// if advertisement is turned off withdraw any previously advertised
+	// cidrs and early return nil.
+	if !params.Enabled {
+		l.Debug("advertisements disabled for BGP instance")
+
+		for _, advrt := range params.CurrentAdvertisements {
+			l.WithField("NLRI", advrt.NLRI.String()).Debug("Withdrawing path")
```
Comment (YutaroHayakawa) [inline pkg/bgpv1/manager/reconcilerv2/advertisements.go]: Let's use log field

File: pkg/bgpv1/manager/reconcilerv2/peer_advertisements_test.go [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+			reqAdvertTypes: []v2alpha1.BGPAdvertisementType{v2alpha1.BGPPodCIDRAdvert},
+			expectedAdverts: map[string]PeerFamilyAdvertisements{
+				"red-peer-65001": map[v2alpha1.CiliumBGPFamily][]v2alpha1.BGPAdvertisement{
+					{Afi: "ipv4", Safi: "unicast"}: nil, // empty advertisement
+				},
+			},
+		},
+	}
+
+	req := require.New(t)
```
Comment (YutaroHayakawa) [inline pkg/bgpv1/manager/reconcilerv2/peer_advertisements_test.go]: Using this req inside the subtest below is invalid. You must create a new req within the t.Run(). 

Per documentation

```
FailNow must be called from the goroutine running the test
```

`require` package calls FailNow internally against given `t` and t.Run spawns a new goroutine to run function. Thus, we shouldn't call FailNow of parent req from subtest.

Same for other tests.

File: pkg/bgpv1/manager/reconcilerv2/peer_advertisements.go [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+
+// GetConfiguredAdvertisements can be called to get all configured advertisements of given BGPAdvertisementType for each peer.
+// Advertisements are selected based on below criteria:
+// Each peer is selected from the BGP node instance configuration. For each peer, the peer configuration is fetched
+// from local store.
+// Peer configuration contains the list of families and the advertisement selector.
+// We iterate over all advertisements ( available from local store ), select only those that match the advertisement
+// selector of the family.
+// Information of peer -> family -> advertisements is returned to the consumer.
+func (p *CiliumPeerAdvertisement) GetConfiguredAdvertisements(conf *v2alpha1.CiliumBGPNodeInstance, selectAdvertTypes ...v2alpha1.BGPAdvertisementType) (PeerAdvertisements, error) {
```
Comment (YutaroHayakawa) [inline pkg/bgpv1/manager/reconcilerv2/peer_advertisements.go:61]: So, the time complexity here is

`peers * families * advertisements` to search selected advertisements
`selectedAdvertisements * advertisement` to search target advertisement types
Then this happens for every advertisement reconciler (PodCIDR or PodIPPool, Service)

At this point we don't know the regular case, so might be a too early optimization, but I think we can pre-populate `advertisement type + family -> advertisements`. So that each reconciler can do an efficient query.

File: pkg/bgpv1/manager/reconcilerv2/advertisements.go [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+	InstanceName string
+
+	CurrentAdvertisements []*types.Path
+	ToAdvertise           []*types.Path
+}
+
+// ExportAdvertisementsReconciler reconciles the state of the BGP advertisements
+// with the provided toAdvertise list and returns a list of the advertisements
+// currently being announced.
+func ExportAdvertisementsReconciler(params *AdvertisementsReconcilerParams) ([]*types.Path, error) {
```
Comment (YutaroHayakawa) [inline pkg/bgpv1/manager/reconcilerv2/advertisements.go]: I'd name this `ReconcileAdvertisement` (verb).

File: pkg/bgpv1/manager/reconcilerv2/advertisements.go [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+	"github.com/sirupsen/logrus"
+
+	"github.com/cilium/cilium/pkg/bgpv1/manager/instance"
+	"github.com/cilium/cilium/pkg/bgpv1/types"
+)
+
+type AdvertisementsReconcilerParams struct {
+	Logger     logrus.FieldLogger
+	Ctx        context.Context
+	Reconciler string
```
Comment (YutaroHayakawa) [inline pkg/bgpv1/manager/reconcilerv2/advertisements.go]: We don't need this parameter because the logger passed from reconcilers already contains the reconciler's name.

File: pkg/bgpv1/manager/reconcilerv2/advertisements.go [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+	Enabled    bool
+
+	Instance     *instance.BGPInstance
+	InstanceName string
```
Comment (YutaroHayakawa) [inline pkg/bgpv1/manager/reconcilerv2/advertisements.go]: Here we already include instance name to the logfield: https://github.com/cilium/cilium/pull/30815/commits/cfd53394f83e379acb9f7344227eab28d4ff200f#diff-4f8b9238ec8c0592396726479067151675de57691dc7e50c09d181befe5055e9R137-R139

We should either remove this field and let reconcilers set the instance name or always rely on AdvertisementReconciler to set the logfield. I think the latter would be more generic and easier to use.

File: pkg/bgpv1/manager/reconcilerv2/advertisements.go [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+
+	"github.com/cilium/cilium/pkg/bgpv1/manager/instance"
+	"github.com/cilium/cilium/pkg/bgpv1/types"
+)
+
+type AdvertisementsReconcilerParams struct {
+	Logger     logrus.FieldLogger
+	Ctx        context.Context
+	Reconciler string
+	Enabled    bool
```
Comment (YutaroHayakawa) [inline pkg/bgpv1/manager/reconcilerv2/advertisements.go]: Can't we express this `Enabled` with zero length `ToAdvertise`?

File: pkg/bgpv1/manager/reconcilerv2/pod_cidr.go [hunk 21]
```diff
@@ -0,0 +1,194 @@
...
+	if err != nil {
+		return err
+	}
+
+	return r.reconcileFamilyAdvertisements(ctx, p, desiredFamilyAdverts)
+}
+
+// getDesiredPathsPerFamily returns a map of desired paths per address family.
+// Note: This returns prefixes per address family. Global routing table will contain prefix per family not per neighbor.
+// Per neighbor advertisement will be controlled by BGP Policy.
```
Comment (YutaroHayakawa) [inline pkg/bgpv1/manager/reconcilerv2/pod_cidr.go:99]: Do we suppose to have this logic in this PR or are you planning to cover it in a separate PR?

File: pkg/bgpv1/manager/reconcilerv2/peer_advertisements_test.go [hunk 22]
```diff
@@ -0,0 +1,254 @@
...
+		name               string
+		peerConfig         []*v2alpha1.CiliumBGPPeerConfig
+		advertisements     []*v2alpha1.CiliumBGPAdvertisement
+		reqAdvertTypes     []v2alpha1.BGPAdvertisementType
+		reqBGPNodeInstance *v2alpha1.CiliumBGPNodeInstance
+		expectedError      bool
+		expectedAdverts    PeerAdvertisements
+	}{
+		{
+			name: "PodCIDR - expecting single advertisement for peer",
```
Comment (YutaroHayakawa) [inline pkg/bgpv1/manager/reconcilerv2/peer_advertisements_test.go]: Are you planning to write tests for other reconciliation types? Otherwise, we can remove this `PodCIDR -` PeerAdvertisements should work for any advertisement types right?

File: pkg/bgpv1/manager/reconcilerv2/peer_advertisements_test.go [hunk 23]
```diff
@@ -0,0 +1,254 @@
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "peer-config-red-podCIDR",
+		},
+		Spec: v2alpha1.CiliumBGPPeerConfigSpec{
+			Families: []v2alpha1.CiliumBGPFamilyWithAdverts{
+				{
+					CiliumBGPFamily: v2alpha1.CiliumBGPFamily{
+						Afi:  "ipv4",
+						Safi: "unicast",
+					},
```
Comment (YutaroHayakawa) [inline pkg/bgpv1/manager/reconcilerv2/peer_advertisements_test.go:144]: I think we should test with dual stack. Currently, the case for multiple address families is not tested.

File: pkg/bgpv1/manager/reconcilerv2/peer_advertisements.go [hunk 24]
```diff
@@ -0,0 +1,138 @@
...
+	for _, peer := range conf.Peers {
+		lp := l.WithField(types.PeerLogField, peer.Name)
+		peerConfig, exist, err := p.peerConfig.GetByKey(resource.Key{Name: peer.PeerConfigRef.Name})
+		if err != nil {
+			return nil, err
+		}
+		if !exist {
+			lp.Debug("Peer config not found, skipping advertisement check")
+			continue
+		}
```
Comment (YutaroHayakawa) [inline pkg/bgpv1/manager/reconcilerv2/peer_advertisements.go:78]: When we hit this case, `GetConfiguredAdvertisements` returns zero advertisements for this peer. It will remove all existing advertisements. We may have a situation that the resource store is not fully synchronized for PeerConfig, but we have BGPInstance. This is possible because the resource store is initialized asynchronously. In that case, we'll delete the route unexpectedly. I guess this will become a problem with GR at agent startup.

File: pkg/bgpv1/manager/reconcilerv2/pod_cidr.go [hunk 25]
```diff
@@ -0,0 +1,194 @@
...
+			agentFamily := types.ToAgentFamily(family)
+
+			// create existingPathsSet for easier existence check.
+			existingPathsSet := sets.New[string]()
+			existingPaths, ok := desiredFamilyAdverts[agentFamily]
+			if ok {
+				for _, existingPath := range existingPaths {
+					existingPathsSet.Insert(existingPath.NLRI.String())
+				}
+			}
```
Comment (YutaroHayakawa) [inline pkg/bgpv1/manager/reconcilerv2/pod_cidr.go]: Looking at the test coverage, this case has not been tested at all.

File: pkg/bgpv1/manager/reconcilerv2/pod_cidr.go [hunk 26]
```diff
@@ -0,0 +1,194 @@
...
+
+			var paths []*types.Path
+			for _, prefix := range desiredPrefixes {
+				path := types.NewPathForPrefix(prefix)
+				path.Family = agentFamily
+
+				// skip if path already exists.
+				if existingPathsSet.Has(path.NLRI.String()) {
+					continue
+				}
```
Comment (YutaroHayakawa) [inline pkg/bgpv1/manager/reconcilerv2/pod_cidr.go]: We don't take path attribute or per-advertisement-type option (e.g., `spec.service`) changes into account here. How do we handle those cases?

File: pkg/bgpv1/manager/reconcilerv2/advertisements.go [hunk 27]
```diff
@@ -0,0 +1,151 @@
...
+
+		key := path.NLRI.String()
+		if m, ok = aset[key]; !ok {
+			aset[key] = &member{
+				configured: true,
+				advrt:      path,
+			}
+			continue
+		}
+		m.configured = true
```
Comment (YutaroHayakawa) [inline pkg/bgpv1/manager/reconcilerv2/advertisements.go]: Do we need this? At this point, member in the `aset` always have `configured: true` right?

File: pkg/bgpv1/manager/reconcilerv2/advertisements.go [hunk 28]
```diff
@@ -0,0 +1,151 @@
...
+		)
+		key := path.NLRI.String()
+		if m, ok = aset[key]; !ok {
+			aset[key] = &member{
+				advertised: true,
+				advrt:      path,
+			}
+			continue
+		}
+		m.advertised = true
```
Comment (YutaroHayakawa) [inline pkg/bgpv1/manager/reconcilerv2/advertisements.go:89]: Looking at the test coverage, we don't test this case at all. Means we don't test the case for `keep` case.

**Review by YutaroHayakawa — CHANGES_REQUESTED** [review]
Thanks for working on this! Did an initial review.

File: pkg/bgpv1/manager/reconcilerv2/pod_cidr.go (hunk 21 above)
Comment (harsimran-pabla) [reply pkg/bgpv1/manager/reconcilerv2/pod_cidr.go:99]: It will be added in subsequent PR - one where we add policy. By default, we will then disable advertisements and only advertise based on the peer policy action.

File: pkg/bgpv1/manager/reconcilerv2/peer_advertisements.go (hunk 16 above)
Comment (harsimran-pabla) [reply pkg/bgpv1/manager/reconcilerv2/peer_advertisements.go:61]: I think it is bit early to optimize. The main factor will be number of peers. Number of advertisements are pretty much constant.

 Linear scan [ Peers ] - O(n) ( number of peers )
 Linear scan [ Families ] - O(m) ( max 2 )
 Linear scan [ Advertisements ] - O(k) ( number of advertisements - 3-4 types, which is again filtered)

Overall complexity will be factor of number of peers. Even if we cache `type + families -> advertisements`, we will need to do linear scan per peer to get correct advertisement. Since number of families and advertisement configurations are expected to be small - I am not very concerned with this approach. 

We can def revisit this if we see it becoming a bottlenech - need some metrics :)

File: pkg/bgpv1/manager/reconcilerv2/peer_advertisements.go (hunk 24 above)
Comment (harsimran-pabla) [reply pkg/bgpv1/manager/reconcilerv2/peer_advertisements.go:78]: I think BGP startup should wait till stores are in sync. This is common problem in all reconcilers based on local store.   It is fixable imo - we need to wait at BGP CP initialization time for stores to get in sync.

File: pkg/bgpv1/manager/reconcilerv2/pod_cidr.go (hunk 26 above)
Comment (harsimran-pabla) [reply pkg/bgpv1/manager/reconcilerv2/pod_cidr.go]: Path attributes will be taken into consideration by policy reconciler.  Service reconciler will take into account additional options field. 

For Pod CIDR reconciler - I believe nothing changes.

File: pkg/bgpv1/manager/reconcilerv2/peer_advertisements_test.go (hunk 22 above)
Comment (harsimran-pabla) [reply pkg/bgpv1/manager/reconcilerv2/peer_advertisements_test.go]: Added more tests - to include v4/v6 families and other advertisement types.

File: pkg/bgpv1/manager/reconcilerv2/pod_cidr.go (hunk 25 above)
Comment (harsimran-pabla) [reply pkg/bgpv1/manager/reconcilerv2/pod_cidr.go]: Added additional test case to have 2 peers. First peer will result in addition of the family. ExistingPathsSet will get populated by second peer.  In the test case, we will get 2 paths for each family ( 4 paths in total ).

File: pkg/bgpv1/manager/reconcilerv2/advertisements.go (hunk 28 above)
Comment (harsimran-pabla) [reply pkg/bgpv1/manager/reconcilerv2/advertisements.go:89]: Modified test case to run the reconciler twice. In which case, we will run into keep case.

**Comment by harsimran-pabla** [comment]
Hi @YutaroHayakawa, I think most of the comments are addressed. There are few open comments, I added my thoughts in those comments. I think most of them are non-blocking. Whenever you get a chance, pls review the changes again. 🙏

File: pkg/bgpv1/manager/reconcilerv2/peer_advertisements.go (hunk 24 above)
Comment (YutaroHayakawa) [reply pkg/bgpv1/manager/reconcilerv2/peer_advertisements.go:78]: Ah, actually, GetByKey returns `ErrStoreUninitialized` error when the store is uninitialized. If we assume the store is already synchronized here, then when we hit the `ErrStoreUninitialized`, it's a bug. We should log it loudly. If we can catch that case in CI (maybe with Error log test?), that would be nice.

Comment (YutaroHayakawa) [reply pkg/bgpv1/manager/reconcilerv2/peer_advertisements.go:78]: The `!exist` case is expected because the user may refer to but forget to create the actual resource. Let's add a test case for this.

File: pkg/bgpv1/gobgp/state.go [hunk 29]
```diff
@@ -141,100 +141,6 @@ func toAgentAfiSafiState(state *gobgp.AfiSafiState) *models.BgpPeerFamilies {
 	return res
 }
 
-// toAgentSessionState translates gobgp session state to cilium bgp session state.
```
Comment (YutaroHayakawa) [inline pkg/bgpv1/gobgp/state.go:144]: Please split refactoring into separate commits.

File: pkg/bgpv1/manager/reconcilerv2/pod_cidr.go [hunk 30]
```diff
@@ -0,0 +1,196 @@
...
+				// we can skip the peer and continue with the next peer.
+				// We do not need to evaluate individual advertisement.
+				continue
+			}
+
+			agentFamily := types.ToAgentFamily(family)
+
+			// create existingPathsSet for easier existence check.
+			existingPathsSet := sets.New[string]()
+			existingPaths, ok := desiredFamilyAdverts[agentFamily]
```
Comment (YutaroHayakawa) [inline pkg/bgpv1/manager/reconcilerv2/pod_cidr.go:115]: Is !ok possible here?

File: pkg/bgpv1/manager/reconcilerv2/pod_cidr.go [hunk 31]
```diff
@@ -0,0 +1,196 @@
...
+	DaemonConfig *option.DaemonConfig
+}
+
+type PodCIDRReconciler struct {
+	logger     logrus.FieldLogger
+	peerAdvert *CiliumPeerAdvertisement
+}
+
+// PodCIDRMetadata is a map of advertisements per family, key is family type
+type PodCIDRMetadata map[types.Family][]*types.Path
```
Comment (YutaroHayakawa) [inline pkg/bgpv1/manager/reconcilerv2/pod_cidr.go]: Let's make this a struct containing this map. The current form is hard to extend.

And nit: PodCIDRMetadata => PodCIDRReconcilerMetadata

File: pkg/bgpv1/manager/reconcilerv2/pod_cidr.go [hunk 32]
```diff
@@ -0,0 +1,196 @@
...
+				path.Family = agentFamily
+
+				// skip if path already exists.
+				if existingPathsSet.Has(path.NLRI.String()) {
+					continue
+				}
+				paths = append(paths, path)
+			}
+
+			desiredFamilyAdverts[types.ToAgentFamily(family)] = append(desiredFamilyAdverts[types.ToAgentFamily(family)], paths...)
```
Comment (YutaroHayakawa) [inline pkg/bgpv1/manager/reconcilerv2/pod_cidr.go]: We already do agent family conversion here. Let's use this. 

https://github.com/cilium/cilium/pull/30815/commits/16b5a4fe58e59c6d566f3e996dabc1b7053d430f#diff-4f8b9238ec8c0592396726479067151675de57691dc7e50c09d181befe5055e9R114

File: pkg/bgpv1/manager/reconcilerv2/pod_cidr.go [hunk 33]
```diff
@@ -0,0 +1,196 @@
...
+		if _, ok := desiredFamilyAdverts[family]; !ok {
+			_, err := ReconcileAdvertisement(&AdvertisementsReconcilerParams{
+				Logger:                l,
+				Ctx:                   ctx,
+				Instance:              p.BGPInstance,
+				CurrentAdvertisements: runningFamilyAdverts[family],
+				ToAdvertise:           nil,
+			})
+			if err != nil {
+				return err
```
Comment (YutaroHayakawa) [inline pkg/bgpv1/manager/reconcilerv2/pod_cidr.go:165]: Reconcile advertisement may fail in the middle of the iteration. However, since we don't record that partial state in runningFamilyAdverts, it is possible to delete the same route twice. GoBGP returns an error in that case. Then we'll get into the failure loop here. I'd change ReconcileAdvertisement to return the current advertisement state even on error, so that we can reflect the actual state to the runningFamilyAdverts.

File: pkg/bgpv1/manager/reconcilerv2/reconcilers.go [hunk 34]
```diff
@@ -32,8 +32,10 @@ type ConfigReconciler interface {
 }
 
 var ConfigReconcilers = cell.ProvidePrivate(
+	NewCiliumPeerAdvertisement,
```
Comment (YutaroHayakawa) [inline pkg/bgpv1/manager/reconcilerv2/reconcilers.go]: Nit: CiliumPeerAdvertisement is not a ConfigReconciler. I'd provide it separately.

File: pkg/bgpv1/types/fake_router.go
Comment (YutaroHayakawa) [inline pkg/bgpv1/types/fake_router.go:1]: Why don't we use a real router?

**Review by YutaroHayakawa — CHANGES_REQUESTED** [review]
Did a second review. Sorry for the delay 🙏

File: pkg/bgpv1/types/fake_router.go
Comment (harsimran-pabla) [reply pkg/bgpv1/types/fake_router.go:1]: I want to decouple gobgp from the reconciler tests. Couple of reasons
- Reconcilers should be agnostic of underlying router implementation.
- We run into port reuse issue, for eg if there is an issue in cleanup of gobgp instance. Rest of the tests will fail to start since port will already be in use.

File: pkg/bgpv1/manager/reconcilerv2/pod_cidr.go (hunk 32 above)
Comment (harsimran-pabla) [reply pkg/bgpv1/manager/reconcilerv2/pod_cidr.go]: good catch

File: pkg/bgpv1/manager/reconcilerv2/pod_cidr.go (hunk 30 above)
Comment (harsimran-pabla) [reply pkg/bgpv1/manager/reconcilerv2/pod_cidr.go:115]: For the first iteration of the peer, 'desiredFamilyAdverts` will be empty. Maybe there is confusion regarding `desiredPeerAdverts` and `desiredFamilyAdverts` ?

File: pkg/bgpv1/manager/reconcilerv2/pod_cidr.go (hunk 33 above)
Comment (harsimran-pabla) [reply pkg/bgpv1/manager/reconcilerv2/pod_cidr.go:165]: Modified Advertisement reconciler. It will return running advertisements in cases where there is an error.

I don't think we have a good way to test this. I will think more about it and add a test case in follow up PR.

**Review by YutaroHayakawa — APPROVED** [review]
Now looks good. Thanks!

**Review by rastislavs — APPROVED** [review]

**Comment by harsimran-pabla** [comment]
[Ingress test failure](https://github.com/cilium/cilium/issues/30281) : https://github.com/cilium/cilium/issues/30281
